### DISCUSSION TITLE
Trailing stop

### DISCUSSION THREAD

**Discussion by ghost** [edtechre/pybroker#172] [Q&A]
How can i add my own trailing stop, for example as the highest close since the entry - ATR, or highest high from entry - x%?

This doesn't work - assume any entry condition is run before this

          entry_price = ctx.close[-1]  # Current price
          ATR = (3.0 * atr_10d[-1])

    elif long_pos and rolling_close[-2] > entry_price and rolling_close[-2] > rolling_close[-3]:
      
      ctx.cancel_stops(ctx.symbol)
      
      ctx.sell_stop_price = rolling_close[-2] - ATR
      ctx.stop_loss = ctx.sell_stop_price

**Comment by ghost** [comment]
ok, good old claude helped me out - leaving here to help out others if useful

          ctx.entry_price = entry_price
          ctx.highest_price = entry_price

    elif long_pos:
      current_price = ctx.close[-1]
      
      # Update highest price if we have a new high
      if not hasattr(ctx, 'highest_price'):
          ctx.highest_price = current_price
      elif current_price > ctx.highest_price:
          ctx.highest_price = current_price
      
      # Calculate trailing stop level (e.g., 2 ATR below the highest price)
      if spy_sma[-1] < spy_sma[-2]:
            trailing_stop = ctx.highest_price - (1.5 * atr_10d[-1])  # Tighter stop (1 ATR)
            stop_type = "Tight"
      else:
          trailing_stop = ctx.highest_price - (3 * atr_10d[-1])  # Normal stop (2 ATR)
          stop_type = "Normal"

      # Check trailing stop condition separately to print message
      trailing_stop_hit = current_price < trailing_stop
      # if trailing_stop_hit:
      #     print(f"Date: {ctx.date} | Symbol: {ctx.symbol} | TRAILING STOP HIT")
      #     print(f"Stop Type: {stop_type} | Highest Price: ${ctx.highest_price:.2f} | Current Price: ${current_price:.2f}")
      #     print(f"Stop Level: ${trailing_stop:.2f} | Distance: ${(ctx.highest_price - trailing_stop):.2f}")
      #     print(f"Profit/Loss: ${(current_price - ctx.entry_price):.2f} | {((current_price - ctx.entry_price) / ctx.entry_price * 100):.2f}%")
      #     print("-" * 80)
      
      # Exit conditions (combine trailing stop with existing exit conditions)
      exit_conditions = [
          #current_price < trailing_stop,  # Trailing stop hit
          trailing_stop_hit
          
      ]
      
      if any(exit_conditions):
          ctx.sell_all_shares()
          # Reset tracking variables
          if hasattr(ctx, 'highest_price'):
              del ctx.highest_price
          if hasattr(ctx, 'entry_price'):
              del ctx.entry_price

**Comment by ghost** [comment]
as above
